File: Try_GameEngine5/src/com/example/try_gameengine/assemble/AssembleViewConfig.java
package com.example.try_gameengine.assemble;

import android.content.Context;
import android.graphics.Bitmap.Config;

public class AssembleViewConfig {
	public enum DirectionConfig{
		NONE, TOP, BOTTOM, LEFT, RIGHT, 
	}
	
	public enum CenterConfig{
		NONE, CENTER, CENTER_HO, CENTER_VIRTICAL
	}
	
	final float x;
	final float y;
	final float w;
	final float h;
	final DirectionConfig directionConfig;
	final CenterConfig centerConfig;
	final float persentX;
	final float persentY;
	
	private AssembleViewConfig(final Builder builder) {
		x = builder.x;
		y = builder.x;
		w = builder.x;
		h = builder.x;
		directionConfig = builder.directionConfig;
		centerConfig = builder.centerConfig;
		persentX = builder.persentX;
		persentY = builder.persentY;
	}
	
	public static AssembleViewConfig createDefault(Context context) {
		return new Builder().build();
	}
	
	public static class Builder {
		private Context context;
		private float x;
		private float y;
		private float w;
		private float h;
		private DirectionConfig directionConfig = DirectionConfig.NONE;
		private CenterConfig centerConfig = CenterConfig.NONE;
		private float persentX;
		private float persentY;
		
		public Builder() {
//			this.context = context.getApplicationContext();
		}
		
		public Builder setXY(float x, float y){
			this.x = x;
			this.y = y;
			return this;
		}
		
		public Builder setWH(float w, float h){
			this.w = w;
			this.h = h;
			return this;
		}
		
		public Builder setDirectionConfig(DirectionConfig directionConfig){
			this.directionConfig = directionConfig;
			return this;
		}
		
		public Builder setCenterConfig(CenterConfig centerConfig){
			this.centerConfig = centerConfig;
			return this;
		}
		
		public Builder setPersentX(float persentX){
			this.persentX = persentX;
			return this;
		}
		
		public Builder setPersentY(float persentY){
			this.persentY = persentY;
			return this;
		}
		
		public AssembleViewConfig build() {
			initEmptyFieldsWithDefaultValues();
			return new AssembleViewConfig(this);
		}
		
		private void initEmptyFieldsWithDefaultValues() {
//			if (taskExecutor == null) {
//				taskExecutor = DefaultConfigurationFactory
//						.createExecutor(threadPoolSize, threadPriority, tasksProcessingType);
//			} else {
//				customExecutor = true;
//			}
//			if (taskExecutorForCachedImages == null) {
//				taskExecutorForCachedImages = DefaultConfigurationFactory
//						.createExecutor(threadPoolSize, threadPriority, tasksProcessingType);
//			} else {
//				customExecutorForCachedImages = true;
//			}
//			if (diskCache == null) {
//				if (diskCacheFileNameGenerator == null) {
//					diskCacheFileNameGenerator = DefaultConfigurationFactory.createFileNameGenerator();
//				}
//				diskCache = DefaultConfigurationFactory
//						.createDiskCache(context, diskCacheFileNameGenerator, diskCacheSize, diskCacheFileCount);
//			}
//			if (memoryCache == null) {
//				memoryCache = DefaultConfigurationFactory.createMemoryCache(memoryCacheSize);
//			}
//			if (denyCacheImageMultipleSizesInMemory) {
//				memoryCache = new FuzzyKeyMemoryCache(memoryCache, MemoryCacheUtils.createFuzzyKeyComparator());
//			}
//			if (downloader == null) {
//				downloader = DefaultConfigurationFactory.createImageDownloader(context);
//			}
//			if (decoder == null) {
//				decoder = DefaultConfigurationFactory.createImageDecoder(writeLogs);
//			}
//			if (defaultDisplayImageOptions == null) {
//				defaultDisplayImageOptions = DisplayImageOptions.createSimple();
//			}
		}

	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/assemble/AssembleViewConfig.kt
package com.example.try_gameengine.assemble

import android.content.Context

class AssembleViewConfig private constructor(builder: Builder) {
    enum class DirectionConfig {
        NONE, TOP, BOTTOM, LEFT, RIGHT,
    }

    enum class CenterConfig {
        NONE, CENTER, CENTER_HO, CENTER_VIRTICAL
    }

    val x: Float
    val y: Float
    val w: Float
    val h: Float
    val directionConfig: DirectionConfig?
    val centerConfig: CenterConfig?
    val persentX: Float
    val persentY: Float

    init {
        x = builder.x
        y = builder.x
        w = builder.x
        h = builder.x
        directionConfig = builder.directionConfig
        centerConfig = builder.centerConfig
        persentX = builder.persentX
        persentY = builder.persentY
    }

    class Builder {
        val context: Context? = null
        var x = 0f
        var y = 0f
        var w = 0f
        var h = 0f
        var directionConfig: DirectionConfig? = DirectionConfig.NONE
        var centerConfig: CenterConfig? = CenterConfig.NONE
        var persentX = 0f
        var persentY = 0f

        fun setXY(x: Float, y: Float): Builder {
            this.x = x
            this.y = y
            return this
        }

        fun setWH(w: Float, h: Float): Builder {
            this.w = w
            this.h = h
            return this
        }

        fun setDirectionConfig(directionConfig: DirectionConfig?): Builder {
            this.directionConfig = directionConfig
            return this
        }

        fun setCenterConfig(centerConfig: CenterConfig?): Builder {
            this.centerConfig = centerConfig
            return this
        }

        fun setPersentX(persentX: Float): Builder {
            this.persentX = persentX
            return this
        }

        fun setPersentY(persentY: Float): Builder {
            this.persentY = persentY
            return this
        }

        fun build(): AssembleViewConfig {
            initEmptyFieldsWithDefaultValues()
            return AssembleViewConfig(this)
        }

        private fun initEmptyFieldsWithDefaultValues() {
//			if (taskExecutor == null) {
//				taskExecutor = DefaultConfigurationFactory
//						.createExecutor(threadPoolSize, threadPriority, tasksProcessingType);
//			} else {
//				customExecutor = true;
//			}
//			if (taskExecutorForCachedImages == null) {
//				taskExecutorForCachedImages = DefaultConfigurationFactory
//						.createExecutor(threadPoolSize, threadPriority, tasksProcessingType);
//			} else {
//				customExecutorForCachedImages = true;
//			}
//			if (diskCache == null) {
//				if (diskCacheFileNameGenerator == null) {
//					diskCacheFileNameGenerator = DefaultConfigurationFactory.createFileNameGenerator();
//				}
//				diskCache = DefaultConfigurationFactory
//						.createDiskCache(context, diskCacheFileNameGenerator, diskCacheSize, diskCacheFileCount);
//			}
//			if (memoryCache == null) {
//				memoryCache = DefaultConfigurationFactory.createMemoryCache(memoryCacheSize);
//			}
//			if (denyCacheImageMultipleSizesInMemory) {
//				memoryCache = new FuzzyKeyMemoryCache(memoryCache, MemoryCacheUtils.createFuzzyKeyComparator());
//			}
//			if (downloader == null) {
//				downloader = DefaultConfigurationFactory.createImageDownloader(context);
//			}
//			if (decoder == null) {
//				decoder = DefaultConfigurationFactory.createImageDecoder(writeLogs);
//			}
//			if (defaultDisplayImageOptions == null) {
//				defaultDisplayImageOptions = DisplayImageOptions.createSimple();
//			}
        }
    }

    companion object {
        fun createDefault(context: Context?): AssembleViewConfig {
            return Builder().build()
        }
    }
}
